### ISSUE TITLE
JSON Parse so slow when running long time

### ISSUE THREAD

**Issue by mrnonz** [jruby/jruby#5993]
### Environment

Provide at least:
* jruby 9.2.9.0 (2.5.7) 2019-10-30 458ad3e OpenJDK 64-Bit Server VM 25.212-b04 on 1.8.0_212-b04 +jit [linux-x86_64]
* Linux 0b518c1ae90f 4.15.0-1055-azure #60-Ubuntu SMP Thu Aug 8 18:29:07 UTC 2019 x86_64 Linux
* I using jruby docker image base on `alpine`

Other relevant info you may wish to add:
* rails 4.2.11

### Expected Behavior

* It's shouldn't consume more time when running process by long time. So should keep more stable than this graph

### Actual Behavior

* When running long time it's consume more and more time (growth) with `to_json` function. (Line no. 9 in my code)

Here graph from newrelic.

<img width="819" alt="Screen Shot 2562-12-09 at 11 49 00" src="https://user-images.githubusercontent.com/9113726/70408235-f4e16a80-1a79-11ea-8cbf-ce1f31e24ca1.png">

and Here my breakdown tx table

<img width="769" alt="Screen Shot 2562-12-09 at 11 50 16" src="https://user-images.githubusercontent.com/9113726/70408278-21958200-1a7a-11ea-8bae-22324a7c25e5.png">

I tracing consume time with this code

```
 ::NewRelic::Agent::MethodTracer.trace_execution_scoped(['Custom/GrapeLoggingMiddlewareRequestLogger#after-json-payload']) do
    json_payload = GrapeLoggingPayload.new(payload).to_json
  end
```

and this is `GrapeLoggingPayload` class

```
class GrapeLoggingPayload
  attr_reader :payload

  def initialize(payload)
    @payload = payload
  end

  def to_json
    to_hash.to_json  # << So slow line
  end

  private

  def to_hash
    id, email = find_person

    {
      method:     payload[:method],
      path:       payload[:path],
      format:     payload[:format].presence,
      status:     payload[:status],
      duration:   payload[:time][:total],
      view:       payload[:time][:view],
      db:         payload[:time][:db],
      remote_ip:  xx,
      xx:         xx,
      user_id:    id,
      user_email: email,
      headers:    format_headers,
      params:     format_params,
      response:   format_response
    }.compact
  end
```

**Comment by enebo** [comment]
@mrnonz just to make sure...You have isolated the timings of find_person and whatever else is happening during to_hash to isolate that Hash#to_json is the thing slowing down?

**Comment by mrnonz** [comment]
@enebo I have already test another part of code and It's depend on how large of object which use to convert from hash to json.

I try to comment out headers, params and response. It's can reduced so much time for convert object but it's still slow bit more by time (growth).

**Comment by enebo** [comment]
@mrnonz ok.  I guess we will try some long json running bench and see if it slows down over time?  Hopefully it will just pop out as obvious.

**Comment by headius** [comment]
This looks like a linear slowdown of the sort caused by some linear-searched list growing without bounds. It should show up in just about any profiling utility, or we should see such a list growing in a heap dump.

@mrnonz If you can provide a simple script for us that reproduces it, we can investigate further. Otherwise I recommend you look at a heap dump or profile yourself.

You can get a sampled profile running JRuby on any Java 8 JDK by passing "--sample" to JRuby. Note that Rails double-launches, and the sample does not show up until exit, so you may have to play with flags and environment variables a bit.

You can also connect to a running JDK using VisualVM and view a current heap dump at any time. I would recommend running the questionable code until it is several times slower than it should be and then pulling a heap dump. You can zip up that file and post it online for us, or just look around in it for anything that appears to be growing large.

Feel free to stop by our Matrix channel at https://matrix.to/#/#jruby:matrix.org if you want to discuss this.

**Comment by mrnonz** [comment]
@headius I will try to profiling my service and heap dump.

thanks

**Comment by mrnonz** [comment]
@enebo @headius I have already try to dump heap and here result.

https://drive.google.com/file/d/1beDSyKhp0xci6Ua5JDk68bTWsKliuSKv/view?usp=sharing

**Comment by headius** [comment]
Note that this is a more narrow reproduction of the issue reported in #5890. If we fix this it's likely we'll fix #5890, so I've closed it in favor of this one.

**Comment by enebo** [comment]
@headius are you sure this is the same cause?  In the heap dump I do not see an unusual amount of methods and I do see a lot of hash map data.

**Comment by headius** [comment]
@enebo Maybe I'm being too optimistic?

What were you expecting to see, relating to #5890?

When I look in the provided heap dump for RubyClass, I see this:

![image](https://user-images.githubusercontent.com/10135/71930304-009b7c00-3161-11ea-8bcd-5d344c267a7d.png)

So that's a similar count to the other issue. An OQL query of all RubyClass with a null basename (anonymous classes, usually singletons) has a count near 2000.

So maybe 7000 RubyClass instances is a lot, but that's more like 5000 named classes plus 2000 singletons, which may not be unreasonable for a very large application.

What hash map data do you mean?

**Comment by headius** [comment]
FWIW the "quick estimate" of retained size for those classes is 101MB, but it's hard to isolate that since every class has a reference back to org.jruby.Ruby and such.

**Comment by enebo** [comment]
@headius Maybe you are or maybe not...My comment about hash was RubyHashEntry is about 18M alone.  Comparing that to redmine it is probably 3x larger but then again this was a JSON app perf issue.  Growing data would be a natural vector of slowing requests.

Looking at your clue and I see 3375 IRClassBody  and 312 IRMetaClassBody.  So not nearly the 6k you are seeing...but how about this...There are 6400 IREvalScripts.  Perusing and almost all of them are:

```ruby
          mod.module_eval <<-RUBY, __FILE__, __LINE__ + 1
            #{defn}
              #{target}
            end
          RUBY
```

in atrribute_methods or active_model.  Looking at how it is rooted it I can see 3 things holding on to it.   staticscope and interpretercontext seem to be 2 (and that makes sense if they are actually executing) but I also see lexicalParent which really should not be needed.  I cannot fully wrap my head around this atm but I do not think any of these should exist.

**Comment by enebo** [comment]
I should add does this application really have 6500 attributes in the database schema?  This makes me wonder if we are not continuously generating something and somehow never freeing it.

**Comment by headius** [comment]
@enebo I guess I'm not sure either issue is ideal; the previous one was just a general "there's a lot of memory being used" and this one at least has a clue that "json is slowing down". So I figure we'll go forward with investigating here and if it turns out not to be json we can open a new issue or modify this one.

**Comment by enebo** [comment]
@headius yeah I am not really sure if they are related or not at this point but we can also change/close based on how the other issue reports.  We may realize they have 6000 evals too or something which comes to light.

**Comment by enebo** [comment]
I noticed redmine also have 2000 IREvalScript instances.  Is it an issue?  eval gets its own var scope right so a closure returned from an eval could require the eval scope lives.  Otherwise, we should not see these things any more.

**Comment by headius** [comment]
@enebo I don't know if it's a problem but it definitely seems like an opportunity. There should be very little an eval'ed method would need from the surrounding eval scope.

**Comment by headius** [comment]
This is a very old issue at this point, and the json library has been practically rewritten since then (and much faster). If there's still an issue, please reopen with a modern reproduction.
